### ISSUE TITLE
[optimization] maybe construct LearningRateSchedule instead of a server_lr_schedule callback?

### ISSUE THREAD

**Issue by amitport** [google-research/federated#5]
https://github.com/google-research/federated/blob/cd6a081f9eb96ced2884c2e58a4b23ed9d9540fa/optimization/shared/fed_avg_schedule.py#L290

Is there a specific reason not to use the optimizer's ability to receive `LearningRateSchedule` as a parameter? `LearningRateSchedule` can be constructed from the command line parameters, and the code that handles `server_lr_schedule` could be removed. (the learning rate is calculated from the optimizer `iter` variable that is being stored in the server state and restored each round)

I understand in the client this is a problem, but why is this behavior included in the server? Am I missing something here?

thank you!

**Comment by nightldj** [comment]
amitport@, a minor comment on the format. Could you try add [optimization] in the issue title so that it is easier to find people could help? Thanks!

**Comment by zcharles8** [comment]
Hi @amitport. This is definitely a way that we could go. However, it's beneficial in `optimizer_utils` to have the behavior be symmetric. That way we don't have to enforce different learning rate schedule behavior depending on whether the optimizer is at `CLIENTS` or at `SERVER`.

Is there a specific issue you're facing related to the server learning rate schedule?

**Comment by amitport** [comment]
@zcharles8 I'm running both federated averaging with a custom aggregation process but I also want to use decaying learning rates.

I'm running code similar to `optimization/main/federated_trainer.py` but using `tff.learning.build_federated_averaging_process` instead of `fed_avg_schedule` and try to add adaptive schedule on top of that.
I've noticed there is a custom federated server schedule code and wondered if there is a need to use it. 

So this is not exactly an issue, but I'm not sure where to open such discussion. This could be a feature request to add aggregation process support to the `fed_avg_schedule` or the other way around, to add client schedule to `build_federated_averaging_process`. (If I know there is interest I may submit a PR)

In any case, I'm not sure what you mean by 
> That way we don't have to enforce different learning rate schedule behavior depending on whether the optimizer is at CLIENTS or at SERVER.

The same command line arguments could be mapped to custom federated code at the client, and only build a simple `LearningRateSchedule` in the server

**Comment by zcharles8** [comment]
@amitport This sounds like desired functionality for `build_federated_averaging_process` if I'm not mistaken. As you correctly point out, we can simply use existing Keras functionality to do server LR decay, but not client LR decay. In general, `fed_avg_schedule` is not intended to be a comprehensive iterative process builder that supports things like custom aggregation processes, so I would recommend instead filing a feature request on TFF. Does that sound like a good course of action?

**Comment by amitport** [comment]
Yes, I'll do that. Thank you
